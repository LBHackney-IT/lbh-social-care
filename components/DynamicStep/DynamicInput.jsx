import PropTypes from 'prop-types';

import * as Inputs from 'components/Form';

const DynamicInput = ({
  component,
  register,
  control,
  id,
  name,
  errors,
  multiStepIndex,
  ...otherProps
}) => {
  const inputName =
    typeof multiStepIndex === 'number'
      ? `${id}[${multiStepIndex}].${name}`
      : name;
  const Component = Inputs[component];
  const sharedProps = {
    name: inputName,
    error: errors[inputName],
    required: otherProps?.rules?.required,
    ...otherProps,
  };
  switch (component) {
    case 'AddressLookup':
    case 'DateInput':
      return <Component control={control} {...sharedProps} />;
    default:
      return (
        <Component register={register(otherProps.rules)} {...sharedProps} />
      );
  }
};

DynamicInput.propTypes = {
  component: PropTypes.string.isRequired,
  register: PropTypes.func.isRequired,
  control: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  rules: PropTypes.shape({
    required: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
  multiStepIndex: PropTypes.number,
};

export default DynamicInput;
