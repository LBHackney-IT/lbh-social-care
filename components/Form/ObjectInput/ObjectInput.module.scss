.inlineInputs {
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
  display: grid;
  column-gap: 1rem;
  align-items: flex-end;
  @media (min-width: 40.0625em) {
    & > div {
      margin-bottom: 0;
    }
  }
}
