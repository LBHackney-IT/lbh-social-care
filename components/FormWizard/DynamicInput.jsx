import PropTypes from 'prop-types';
import get from 'lodash/get';

import * as Inputs from 'components/Form';

const enhanceValidate = (validate, currentData) =>
  typeof validate === 'function'
    ? (value) => validate(value, currentData)
    : Object.entries(validate).reduce(
        (acc, [key, validation]) => ({
          ...acc,
          [key]: (value) => validation(value, currentData),
        }),
        {}
      );

const DynamicInput = (props) => {
  const {
    component,
    control,
    errors,
    name,
    options,
    currentData,
    register,
    rules,
    ...otherProps
  } = props;
  const Component = Inputs[component];
  if (!Component)
    throw new Error(`"${component}" is not a supported component type.`);
  const sharedProps = {
    name,
    error: get(errors, name),
    required: rules?.required,
    options: typeof options === 'function' ? options(currentData) : options,
    rules: {
      ...rules,
      validate: rules?.validate && enhanceValidate(rules.validate, currentData),
    },
    ...otherProps,
  };
  if (typeof options === 'function' && !sharedProps.options) {
    return null;
  }
  switch (component) {
    case 'ObjectInput':
      return <Component {...props} />;
    case 'AddressLookup':
    case 'Autocomplete':
    case 'DateInput':
      return <Component control={control} {...sharedProps} />;
    default:
      return <Component register={register} {...sharedProps} />;
  }
};

DynamicInput.propTypes = {
  component: PropTypes.string.isRequired,
  control: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  currentData: PropTypes.object.isRequired,
  labelSize: PropTypes.string,
  rules: PropTypes.shape(),
  register: PropTypes.func,
  options: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
};

export default DynamicInput;
