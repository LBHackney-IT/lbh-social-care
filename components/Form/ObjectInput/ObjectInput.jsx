import PropTypes from 'prop-types';
import cx from 'classnames';

import DynamicInput from 'components/FormWizard/DynamicInput';

import styles from './ObjectInput.module.scss';

const ObjectInput = ({
  label,
  hint,
  name,
  labelSize = 'm',
  isInline,
  components,
  ...otherProps
}) => (
  <div className="govuk-form-group govuk-!-margin-bottom-0">
    <label className={`govuk-label govuk-label--${labelSize}`}>{label}</label>
    {hint && (
      <span id={`${name}-hint`} className="govuk-hint">
        {hint}
      </span>
    )}
    <div className={cx({ [styles.inlineInputs]: isInline })}>
      {components.map((subComponent) => (
        <DynamicInput
          key={subComponent.name}
          {...otherProps}
          {...subComponent}
          labelSize="s"
          name={`${name}.${subComponent.name}`}
        />
      ))}
    </div>
  </div>
);

ObjectInput.propTypes = {
  label: PropTypes.string,
  labelSize: PropTypes.oneOf(['s', 'm', 'l', 'xl']),
  isInline: PropTypes.bool,
  hint: PropTypes.string,
  name: PropTypes.string.isRequired,
  components: PropTypes.array.isRequired,
};

export default ObjectInput;
