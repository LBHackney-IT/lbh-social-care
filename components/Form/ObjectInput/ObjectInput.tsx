import cx from 'classnames';
import { Control, ErrorOption } from 'react-hook-form';

import { ObjectInput as Props } from 'components/Form/types';
import DynamicInput from 'components/FormWizard/DynamicInput';

import styles from './ObjectInput.module.scss';

export interface DynamicComponent extends Props {
  currentData: Record<string, unknown>;
  control: Control;
  errors: ErrorOption;
}

const ObjectInput = ({
  label,
  hint,
  name,
  labelSize = 'm',
  isInline,
  // this props is used for UI purpose in the summary
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  summaryInline,
  components,
  ...otherProps
}: DynamicComponent): React.ReactElement => (
  <div className="govuk-form-group govuk-!-margin-bottom-0">
    <label className={`lbh-label govuk-label--${labelSize}`}>{label}</label>
    {hint && (
      <span id={`${name}-hint`} className="govuk-hint">
        {hint}
      </span>
    )}
    <div className={cx({ [styles.inlineInputs]: isInline })}>
      {components.map((subComponent) => (
        <DynamicInput
          key={subComponent.name}
          {...otherProps}
          {...subComponent}
          labelSize="s"
          name={`${name}.${subComponent.name}`}
        />
      ))}
    </div>
  </div>
);

export default ObjectInput;
