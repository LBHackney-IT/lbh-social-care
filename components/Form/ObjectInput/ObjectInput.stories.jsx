import ObjectInput from './ObjectInput';

import { useForm } from 'react-hook-form';
import Button from 'components/Button/Button';

export default {
  title: 'Form Components/ObjectInput',
  component: ObjectInput,
};

const Template = (args) => {
  const { register, errors, handleSubmit } = useForm();
  const onSubmit = (data) => {
    alert(JSON.stringify(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <ObjectInput
        name="phone_number"
        label="Phone Number"
        errors={errors}
        components={[
          {
            component: 'PhoneInput',
            name: 'phoneNumber',
            label: 'Phone number',
            rules: { required: true },
          },
          {
            component: 'TextInput',
            name: 'phoneType',
            label: 'Phone type',
          },
        ]}
        register={register}
        {...args}
      />
      <Button className="govuk-button" label="Next" type="submit" />
    </form>
  );
};

export const Default = Template.bind({});

export const InLine = Template.bind({});
InLine.args = {
  isInline: true,
};

export const MultiComponent = Template.bind({});
MultiComponent.args = {
  isInline: true,
  components: [
    {
      component: 'TextInput',
      name: 'first',
      label: 'First',
    },
    {
      component: 'TextInput',
      name: 'second',
      label: 'Second',
    },
    {
      component: 'TextInput',
      name: 'third',
      label: 'Third',
    },
    {
      component: 'TextInput',
      name: 'fourth',
      label: 'Fourth',
    },
    {
      component: 'TextInput',
      name: 'fifth',
      label: 'Fifth',
    },
  ],
};
