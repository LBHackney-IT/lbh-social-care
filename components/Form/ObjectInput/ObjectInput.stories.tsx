import { Story } from '@storybook/react';

import ObjectInput, { DynamicComponent } from './ObjectInput';

import { useForm } from 'react-hook-form';
import Button from 'components/Button/Button';

export default {
  title: 'Form Components/ObjectInput',
  component: ObjectInput,
};

const Template: Story<DynamicComponent> = (args) => {
  const { register, errors, handleSubmit } = useForm();
  const onSubmit = (data: Record<string, unknown>) => {
    alert(JSON.stringify(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <ObjectInput
        label="Phone Number"
        register={register}
        {...args}
        name="phone_number"
        errors={errors}
      />
      <Button className="govuk-button" label="Next" type="submit" />
    </form>
  );
};

export const Default = Template.bind({});
Default.args = {
  components: [
    {
      component: 'PhoneInput',
      name: 'phoneNumber',
      label: 'Phone number',
      rules: { required: true },
    },
    {
      component: 'TextInput',
      name: 'phoneType',
      label: 'Phone type',
    },
  ],
};

export const InLine = Template.bind({});
InLine.args = {
  ...Default.args,
  isInline: true,
};

export const MultiComponent = Template.bind({});
MultiComponent.args = {
  isInline: true,
  components: [
    {
      component: 'TextInput',
      name: 'first',
      label: 'First',
    },
    {
      component: 'TextInput',
      name: 'second',
      label: 'Second',
    },
    {
      component: 'TextInput',
      name: 'third',
      label: 'Third',
    },
    {
      component: 'TextInput',
      name: 'fourth',
      label: 'Fourth',
    },
    {
      component: 'TextInput',
      name: 'fifth',
      label: 'Fifth',
    },
  ],
};
