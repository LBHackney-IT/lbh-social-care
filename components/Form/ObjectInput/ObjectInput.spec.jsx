import { render } from '@testing-library/react';

import ObjectInput from './ObjectInput';

import DynamicInput from 'components/FormWizard/DynamicInput';

jest.mock('components/FormWizard/DynamicInput', () =>
  jest.fn(() => 'MockedDynamicInput')
);

describe('ObjectInput', () => {
  const props = {
    label: 'i am a object input',
    hint: 'i accept multiple inputs',
    name: 'oo',
    isInline: true,
    components: [
      {
        component: 'TextInput',
        name: 'first',
      },
      {
        component: 'TextInput',
        name: 'second',
      },
      {
        component: 'TextInput',
        name: 'third',
      },
    ],
    foo: 'bar',
  };

  it('should render properly', () => {
    const { asFragment } = render(<ObjectInput {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should pass the correct props to DynamicInput', () => {
    render(<ObjectInput {...props} />);
    expect(DynamicInput).toHaveBeenCalled();
    expect(DynamicInput).toHaveBeenNthCalledWith(
      1,
      {
        component: 'TextInput',
        foo: 'bar',
        labelSize: 's',
        name: 'oo.first',
      },
      {}
    );
    expect(DynamicInput).toHaveBeenNthCalledWith(
      2,
      {
        component: 'TextInput',
        foo: 'bar',
        labelSize: 's',
        name: 'oo.second',
      },
      {}
    );
    expect(DynamicInput).toHaveBeenNthCalledWith(
      3,
      {
        component: 'TextInput',
        foo: 'bar',
        labelSize: 's',
        name: 'oo.third',
      },
      {}
    );
  });
});
