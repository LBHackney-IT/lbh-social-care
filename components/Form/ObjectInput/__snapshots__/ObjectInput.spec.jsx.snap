// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInput should render properly 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group govuk-!-margin-bottom-0"
  >
    <label
      class="lbh-label govuk-label--m"
    >
      i am a object input
    </label>
    <span
      class="govuk-hint"
      id="oo-hint"
    >
      i accept multiple inputs
    </span>
    <div
      class="inlineInputs"
    >
      MockedDynamicInputMockedDynamicInputMockedDynamicInput
    </div>
  </div>
</DocumentFragment>
`;
